implement remote validator acceptance test

rolfyone opened the pull request:
- Shows attestation, aggregation, block production occurring from a remote validator node, and the beacon node not having ownership of any validators.

fixes #3247

Signed-off-by: Paul Harris <paul.harris@consensys.net>

## Documentation

- [X] I thought about documentation and added the `documentation` label to this PR if updates are required.

rolfyone committed:
implement remote validator acceptance test

 - Shows attestation, aggregation, block production occurring from a remote validator node, and the beacon node not having ownership of any validators.

fixes #3247

Signed-off-by: Paul Harris <paul.harris@consensys.net>

rolfyone committed:
Merge remote-tracking branch 'upstream/master' into 3247

rolfyone committed:
Merge remote-tracking branch 'upstream/master' into 3247

ajsutton commented on acceptance-tests/src/acceptance-test/java/tech/pegasys/teku/test/acceptance/RemoteValidatorAcceptanceTest.java:
```diff
@@ -0,0 +1,56 @@
+public class RemoteValidatorAcceptanceTest extends AcceptanceTestBase {
+  @Test
+  void shouldCreateAttestationsWithRemoteValidator() throws Exception {
+    final int VALIDATOR_COUNT = 8;
```
nit: Either need to make the name camel case or make this static constant at the top of the file.

ajsutton commented on acceptance-tests/src/acceptance-test/java/tech/pegasys/teku/test/acceptance/RemoteValidatorAcceptanceTest.java:
```diff
@@ -0,0 +1,56 @@
+  @Test
+  void shouldCreateAttestationsWithRemoteValidator() throws Exception {
+    final int VALIDATOR_COUNT = 8;
+    final TekuNode node1 =
```
nit: I'd probably call this `beaconNode` and `validatorNode1`  `validatorClient` since that's the typical terms we'd use for this kind of setup.

ajsutton commented on acceptance-tests/src/acceptance-test/java/tech/pegasys/teku/test/acceptance/RemoteValidatorAcceptanceTest.java:
```diff
@@ -0,0 +1,56 @@
+    assertThat(validatorNode1.getFilteredOutput("Published attestation").size())
+        .isGreaterThanOrEqualTo(2);
+    assertThat(validatorNode1.getFilteredOutput("Published block").size())
+        .isGreaterThanOrEqualTo(2);
```
Rather than waiting for 3 slots then checking logs, we should just have a waiter that looks for these logs and retries for some period of time.  It's possible the beacon node can startup faster than the validator node and process 3 slots before the validator even gets started and if we find the log messages earlier we can bail out.

I'd also just look for one of them - if we can successfully publish we're ok.

rolfyone committed:
review feedback

Signed-off-by: Paul Harris <paul.harris@consensys.net>

rolfyone committed:
add test case for validators available before beacon node is up.

Signed-off-by: Paul Harris <paul.harris@consensys.net>

rolfyone committed:
Merge remote-tracking branch 'upstream/master' into 3247

rolfyone committed:
spotless

Signed-off-by: Paul Harris <paul.harris@consensys.net>

ajsutton commented on acceptance-tests/src/testFixtures/java/tech/pegasys/teku/test/acceptance/dsl/TekuNode.java:
```diff
@@ -158,6 +145,10 @@ public void waitForNewBlock() {
 
+  public void waitForSlot(final long slot) {
+    waitFor(() -> assertThat(getFilteredOutput("Slot: " + slot + ",")).isNotEmpty(), 1, MINUTES);
+  }
```
We probably should query the API for this rather than looking for log messages.  Log messages are basically a last resort when we can't find the info by another means (or if we're really specifically after a log message).

I don't think this method is being used anymore anyway.

ajsutton commented on acceptance-tests/src/testFixtures/java/tech/pegasys/teku/test/acceptance/dsl/TekuNode.java:
```diff
@@ -417,8 +414,9 @@ public Config withDepositsFrom(final BesuNode eth1Node) {
+      LOG.debug("rest-api-host-allowlist:" + allowList);
+      configMap.put("rest-api-host-allowlist", allowList);
       return this;
     }
```
I suspect we should just default this to `*` in ATs so we don't have to worry about it.

rolfyone committed:
set the connection retry maximum duration to be the seconds-per-slot time.

Signed-off-by: Paul Harris <paul.harris@consensys.net>

rolfyone committed:
spotless

Signed-off-by: Paul Harris <paul.harris@consensys.net>

rolfyone committed:
initial fix for remote validator starting first, and the trouble that can occur.

before the fix there are no attestations in the `shouldCreateAttestationsWithRemoteValidatorStartingFirst` test, and after, attestations are created from slot 1.

Signed-off-by: Paul Harris <paul.harris@consensys.net>

rolfyone committed:
Merge branch 'master' into 3247

rolfyone committed:
now that the remote validator starting first is working, I can remove the extra message search and extended wait for messages.

Signed-off-by: Paul Harris <paul.harris@consensys.net>

rolfyone committed:
cleanup

Signed-off-by: Paul Harris <paul.harris@consensys.net>

ajsutton commented on acceptance-tests/src/acceptance-test/java/tech/pegasys/teku/test/acceptance/RemoteValidatorAcceptanceTest.java:
```diff
@@ -0,0 +1,73 @@
+                .withNetwork("minimal")
+                .withInteropValidators(0, VALIDATOR_COUNT)
+                .withBeaconNodeEndpoint(beaconRestUrl));
+  }
```
Probably better to do this in a `@BeforeEach` method and make `beaconNode` and `validatorClient` fields.

ajsutton reviewed (approved):
LGTM. Lets just move the validator index fix loading fix into a separate PR.

rolfyone committed:
Revert "initial fix for remote validator starting first, and the trouble that can occur."

This reverts commit de53c205

Signed-off-by: Paul Harris <paul.harris@consensys.net>

rolfyone committed:
review comments.

Signed-off-by: Paul Harris <paul.harris@consensys.net>

rolfyone committed:
Merge remote-tracking branch 'upstream/master' into 3247

rolfyone committed:
spotless

Signed-off-by: Paul Harris <paul.harris@consensys.net>

rolfyone committed:
increase timeout as CI seems to be running slower than local pc.

Signed-off-by: Paul Harris <paul.harris@consensys.net>

rolfyone committed:
Merge branch 'master' into 3247
